File: ios/CardScanner/CardScannerView.m
//
//  CardScannerView.m
//  CardScanner
//
//  Created by Howl on 01/06/2022.
//  Copyright © 2022 Facebook. All rights reserved.
//

#import "CardScannerView.h"

@implementation CardScannerView

- (instancetype)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        NSLog(@"CardScannerView was initialized in init with Frame");
        [self initializeRecognizer];
    }
    return self;
}

- (void) initializeRecognizer {
    UIColor *defaultColor = [UIColor colorWithRed: 0.00 green: 0.48 blue: 1.00 alpha: 1.00];
    _recognizer = [[PayCardsRecognizer alloc] initWithDelegate:self recognizerMode:PayCardsRecognizerDataModeNumber | PayCardsRecognizerDataModeDate | PayCardsRecognizerDataModeName resultMode:PayCardsRecognizerResultModeAsync container:self frameColor:defaultColor isShowLayer:true];
}

- (void)layoutSubviews
{
    [super layoutSubviews];
    [_recognizer startCamera];
}

- (void)setFrameColor:(UIColor *)frameColor
{
    if (frameColor != nil) {
        const CGFloat *components = CGColorGetComponents(frameColor.CGColor);
        CGFloat r = components[1];
        CGFloat g = components[2];
        CGFloat b = components[3];
        CGFloat a = components[0];
        UIColor *convertedColor = [UIColor colorWithRed:r green:g blue:b alpha:a];
        [_recognizer updateFrameColor:convertedColor];
    }else {
        [_recognizer updateFrameColor:[UIColor colorWithRed: 0.00 green: 0.48 blue: 1.00 alpha: 1.00]];
    }
}

- (void)toggleFlash {
    [_recognizer toggleFlash];
}

- (void)resetResult {
    [_recognizer resetResult];
}

#pragma mark PayCardsRecognizerPlatformDelegate

- (void)payCardsRecognizer:(PayCardsRecognizer * _Nonnull)payCardsRecognizer didRecognize:(PayCardsRecognizerResult * _Nonnull)result {
    if(result.isCompleted && self.onDidScanCard) {
        self.onDidScanCard(@{
            @"cardNumber": result.recognizedNumber ?: @"",
            @"expiryMonth": result.recognizedExpireDateMonth ?: @"",
            @"expiryYear": result.recognizedExpireDateYear ?: @"",
            @"holderName": result.recognizedHolderName ?: @"",
        });
    }
}

@end

File: ios/CardScanner/CardScannerView.swift
//
//  CardScannerView.swift
//
//  Created by Howl on 01/06/2022.
//  Copyright © 2022 Facebook. All rights reserved.
//

@objc(CardScannerViewManager)
class CardScannerViewManager: RCTViewManager {
    
    override static func requiresMainQueueSetup() -> Bool {
        return true
    }
    
    override func view() -> UIView {
        return CardScannerView()
    }
    
    @objc func toggleFlash(_ node:NSNumber) {
        DispatchQueue.main.async {
            let view = self.bridge.uiManager.view(forReactTag: node) as! CardScannerView
            view.toggleFlash()
        }
    }
    
    @objc func resetResult(_ node:NSNumber) {
        DispatchQueue.main.async {
            let view = self.bridge.uiManager.view(forReactTag: node) as! CardScannerView
            view.resetResult()
        }
    }
}

class CardScannerView : UIView, PayCardsRecognizerPlatformDelegate {
    private let defaultColor = UIColor(hue: 0.5861, saturation: 1, brightness: 0.99, alpha: 1.0)
    
    private var recognizer : PayCardsRecognizer!
    
    @objc var onDidScanCard: RCTDirectEventBlock?
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        recognizer = PayCardsRecognizer(delegate: self, recognizerMode: [.number, .date, .name], resultMode: .async, container: self, frameColor: defaultColor, isShowLayer: true)
    }
    
    @objc func setFrameColor(_ val: UIColor?) {
        recognizer.updateFrameColor(val ?? defaultColor)
    }
    
    override func layoutSubviews() {
        super.layoutSubviews()
        recognizer.startCamera()
    }
    
    @objc func payCardsRecognizer(_ payCardsRecognizer: PayCardsRecognizer, didRecognize result: PayCardsRecognizerResult) {
        if result.isCompleted {
            if self.onDidScanCard != nil {
                self.onDidScanCard!(["cardNumber": result.recognizedNumber ?? "", "expiryMonth": result.recognizedExpireDateMonth ?? "","expiryYear": result.recognizedExpireDateYear ?? "", "holderName": result.recognizedHolderName ?? "", ])
            }
        }
    }
    
    public func toggleFlash() {
        recognizer.toggleFlash()
    }
    
    public func resetResult() {
        recognizer.resetResult()
    }
    
    required public init?(coder aDecoder: NSCoder) {
        fatalError( "init(coder:) has not been implemented" )
    }
    
}
